src/Ledger/TokenAlgebra/ValueSet.lagda

```diff
@@ -0,0 +1,249 @@
+                                          renaming ( Map to _⇀_ )
+open import Axiom.Set.Rel th              using ( dom )
+
+open import Data.Nat                      using (ℕ ; _≤_ ; _+_)
+open import Data.Nat.Properties           using (+-0-commutativeMonoid)
+
+open import Ledger.Prelude                using ( DecEq )
+open import Ledger.TokenAlgebra.Base      using ( TokenAlgebra ; Quantity)
+
+open import Prelude                       using ( _,_ ; _×_ ; module Equivalence; _∘_ ; _≗_ ; id )
+open import Prelude                       using ( _≡_ ; refl ; 0ℓ; trans ; sym)
+
+open import Relation.Binary               using (IsEquivalence )
+
+open Theory th                            using (_∈_ ; spec-∈ ; singleton ; ∈-singleton)
+                                          renaming (map to mapˢ ; Set to Set' )
+open import Relation.Binary.PropositionalEquality.Core using ( module ≡-Reasoning )
+import Relation.Binary.Core  as stdlib
+
+\end{code}
+
+\subsubsection{Derived types}
+
+(See Fig 3 of the
+\href{https://github.com/input-output-hk/cardano-ledger/releases/latest/download/mary-ledger.pdf}%
+{Mary ledger specification}.)
+
+\begin{itemize}
+\item \AgdaBound{AssetName} is a byte string used to distinguish different assets with the same \AgdaBound{PolicyId}.
+\item \AgdaBound{AssetId} is a product type consisting of a \AgdaBound{PolicyId} and an \AgdaBound{AssetName}.
+\item \AgdaBound{AdaId} is the Id for the asset Ada.
+\item \AgdaBound{Quantity} is the type of amounts of assets.
+\end{itemize}
+
+In the formal ledger specification \AgdaBound{AssetId} is sometimes viewed as a direct sum type,
+the inhabitants of which belong to either \AgdaBound{AdaIdType} or the product
+\AgdaBound{PolicyId}~\AgdaBound{×}~\AgdaBound{AssetName}; if we were adhering to that point of view,
+then we would have defined
+\AgdaBound{AssetId}
+  = \AgdaBound{AdaIdType}~\AgdaBound{⊎}~(\AgdaBound{PolicyId}~\AgdaBound{×}~\AgdaBound{AssetName}).
+
+Finally, we define a record type with a single inhabitant with which we may wish to
+represent the type of Ada (rather than viewing Ada as just another asset).
+
+\begin{code}
+record AdaIdType : Type where
+  instance constructor AdaId
+\end{code}
+
+
+\subsection{Definition of the value monoid}
+
+An inhabitant of `Value` is a map denoting a finite collection of quantities of assets.
+
+\begin{code}
+open TokenAlgebra
+open CommutativeMonoid renaming (_∙_ to _⋆_) hiding (refl ; sym ; trans)
+open Algebra
+
+
+Value-TokenAlgebra :  (assetNameType  : Type                                       )
+                      (isTotal        : ∀ x {m : (PolicyId × assetNameType) ⇀ Quantity} → x ∈ dom (m ˢ))
+                      (specialPolicy  : PolicyId                                   )
+                      (specialAsset   : assetNameType                              )
+                      (decEqVal       : DecEq ((PolicyId × assetNameType) ⇀ Quantity))
```

Comment:
We'll get in trouble with this one, because propositional equality isn't the correct one here. We probably should change `DecEq` in `TokenAlgebra` to decidability of the equivalence relation on the monoid.

Reply:
Yeah, I wasn't sure what we actually needed to assume was decidable.  Presumably, we want to assume as little as possible.  I'll try to follow your advice here and change `DecEq` in `TokenAlgebra` to decidability of the equivalence relation on the monoid.  Thanks. 👍🏼